frog/imports/api/sessions.js

```diff
@@ -129,16 +132,29 @@ Meteor.methods({
       });
     });
     setTeacherSession(sessionId);
+    return sessionId;
   },
   'flush.session': sessionId => {
-    const session = Sessions.findOne(sessionId)
-    const graphId = session.graphId
-
+    const session = Sessions.findOne(sessionId);
+    if (!session) {
+      return;
+    }
+    const graphId = session.graphId;
+    Sessions.remove(sessionId);
     Graphs.remove(graphId);
     Activities.remove({ graphId });
     Operators.remove({ graphId });
     Connections.remove({ graphId });
-
-    Sessions.remove(sessionId);
+  },
+  'sessions.restart': session => {
+    const graphId = session.fromGraphId;
+    if (!graphId || !session) {
+      return;
+    }
+    Meteor.call('flush.session', session._id);
+    const newSessionId = Meteor.call('add.session', graphId);
+    Meteor.call('session.joinall', newSessionId);
+    runSession(newSessionId);
+    nextActivity(newSessionId);
```

Comment:
I understand it is the fastest way to implement this functionality and it works as a temporary solution.

The two concerns I see with this implementation is 
- The list of subscribed students. We could avoid to call joinAll and only join the already subscribed students
- Having modified the session graph before deciding to restart it; With this implementation you lose these modifications

A different strategy for restarting a session could be:
- Delete saved reactive data and operator's products, and update nodes for being not computed
- Reset session's state, timeInGraph, startedAt, openActivities

In general I believe that restarting a session will not be an often used feature (Do you imagine a professor restarting his lecture after 5 minutes? ). Only us as developers feel the need to do so, I believe.